addon/mixins/callback-route-mixin.js

```diff
@@ -0,0 +1,67 @@
+import Ember from 'ember';
+
+const { inject: { service }, Mixin, isEmpty } = Ember;
+
+export default Mixin.create({
+  /**
+   The session service.
+
+   @property session
+   @readOnly
+   @type SessionService
+   @public
+   */
+  session: service('session'),
+
+  /**
+   The authenticator that should be used to authenticate the callback.
+
+   @property authenticator
+   @type String
+   @default null
+   @public
+   */
+  authenticator: null,
+
+  /**
+   The error that ocurred during the authentication attempt.
+
+   @property error
+   @type String
+   @default null
+   @public
+   */
+  error: null,
+
+  /**
+   The route that should be transitioned to on error.
+
+   @property routeOnError
+   @type String
+   @default null
+   @public
+   */
+  routeOnError: null,
+
+  /**
+   Authenticates the passed authentication server response and sets the error
+   property if the authentication failed.
+
+   If the routeOnError property is set it will transition to the set route with
+   the error code specified as a query parameter.
+
+   @method activate
+   @public
+   */
+  activate() {
+    let authenticator = this.get('authenticator');
+
+    return this.get('session').authenticate(authenticator).catch((err) => {
+      this.set('error', err);
+
+      if (!isEmpty(this.get('routeOnError'))) {
+        return this.transitionTo(this.get('routeOnError'), { queryParams: { error: err } });
```

Comment:
I'd really leave this out and leave it to implementors to redirect on errors by overriding the `activate` method:

```js
activate() {
  this._super(...arguments).catch(() => {
    return this.transitionTo(this.get('routeOnError'), { queryParams: { error: err } });
  });
});
```